subprojects/tooling-api-builders/src/main/java/org/gradle/tooling/internal/provider/runner/ClientForwardingTestOperationListener.java

```diff
@@ -121,6 +121,20 @@ private DefaultTestDescriptor toTestDescriptorForTest(OperationIdentifier buildO
         return new DefaultTestDescriptor(id, name, displayName, testKind, null, className, methodName, parentId, taskPath);
     }
 
+    /**
+     * This method returns a display name which is "compatible with" what previous
+     * Gradle versions did.
+     */
+    private static String backwardsCompatibleDisplayNameOf(TestDescriptorInternal descriptor) {
+        String className = descriptor.getClassName();
+        String methodName = descriptor.getName();
+        String displayName = descriptor.getDisplayName();
+        if (methodName != null && methodName.equals(displayName) || className != null && className.equals(displayName)) {
+            return descriptor.toString();
+        }
+        return displayName;
```

Comment:
Given this is solely computed by information from the descriptor, have you considered moving it there? It feels a bit funny that our descriptor has a concept of a `displayname` but override this in other places with more logic. Wonder if this actually belongs to `getDisplayName`

Reply:
> Given this is solely computed by information from the descriptor, have you considered moving it there?

Yes I did but it felt like it's cleaner to keep the descriptor "as it should be" and only deal with the backwards compatibility for the TAPI side. But I'm happy to be convinced otherwise.